File: leetcode/90-subsetsWithDup.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> list = new ArrayList();
        Arrays.sort(nums);
        backtrack(list, new ArrayList<Integer>(), nums, 0);
        return list;
    }
    private void backtrack(List<List<Integer>> list, List<Integer> item, int[] nums, int start){
        list.add(new ArrayList(item));
        for(int i = start; i < nums.length; i++){
            while(i > start && i < nums.length && nums[i] == nums[i-1])
                i++;
            if(i == nums.length) return;
            item.add(nums[i]);
            backtrack(list, item, nums, i+1);
            item.remove(item.size()-1);
        }
    }
}

File: leetcode/90-subsetsWithDup.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        if(nums.size() == 0) return res;
        vector<int> item;
        sort(nums.begin(), nums.end());
        backtrack(res, item, 0, nums);
        return res;
    }
    void backtrack(vector<vector<int>>& res, vector<int>& item, int start, vector<int>& nums) {
        res.push_back(vector<int>(item));
        for(int i = start; i < nums.size(); i++) {
            if(i > start && nums[i] == nums[i-1])
                continue;
            item.push_back(nums[i]);
            backtrack(res, item, i+1, nums);
            item.pop_back();
        }
    }
};
